js/queues.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * An ordered queue where all the elements are unique, according to the equivalencyCheck.
+ * On insert, the element will only be added if there is not already a higher-priority equivalent element in the queue.
+ * If there is a lower-priority equivalent element in the queue, it will be removed.
+ */
+export class UniquePriorityQueue {
+	constructor(elementMatcher, priorityFunction) {
+		this.first = null;
+		this.elementMatcher = elementMatcher;
+		this.priorityFunction = priorityFunction;
+	}
+
+	push(value) {
```

Comment:
The name of this function is a bit misleading. It suggests that it only inserts an element, but the function might also remove existing elemnts as well. Maybe `update` or `set` would work...

Reply:
Maybe it's my Java background showing through, but since this is a type of queue, I think all queues should have the same (similar at least) methods, and the implementation of those methods is up to them. For example, a Set has an "add" method even though, if the element is already in the set, it won't be added again.

Of course I will change it if you insist!